File: java/Solver.java
import java.util.*;
import java.io.*;

class Solver {
    static String letters;

    static int count(String word, char letter) {
        return (int) word.chars().filter(c -> c == letter).count();
    }

    static boolean valid(String word) {
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);

            if (count(word, c) > count(letters, c)) {
                return false;
            }
        }

        return true;
    }

    static List<String> generateWords() {
        List<String> words = new ArrayList<>();

        try (BufferedReader reader = new BufferedReader(new FileReader("dictionary.txt"))) {
            String line = reader.readLine();

            while (line != null) {
                if (valid(line)) {
                    words.add(line);
                }

                line = reader.readLine();
            }
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return words;
    }

    static Map<Character, Integer> scores = new HashMap<Character, Integer>() {{
        put('a', 1); put('b', 3); put('c', 3); put('d', 2);
        put('e', 1); put('f', 4); put('g', 2); put('h', 4);
        put('i', 1); put('j', 8); put('k', 5); put('l', 1);
        put('m', 3); put('n', 1); put('o', 1); put('p', 3);
        put('q', 10); put('r', 1); put('s', 1); put('t', 1);
        put('u', 1); put('v', 4); put('w', 4); put('x', 8);
        put('y', 4); put('z', 10);
    }};

    static int score(String word) {
        int result = 0;

        for (int i = 0; i < word.length(); i++) {
            result += scores.get(word.charAt(i));
        }

        return result;
    }

    static List<String> sortWords(List<String> words) {
        words.sort(new Comparator<String>() {
            @Override
            public int compare(String first, String second) {
                return score(first) - score(second);
            }
        });

        return words;
    }

    static void displayWords(List<String> words) {
        int length = words.size();

        for (int i = length; i > 0; i--) {
            System.out.println(i + ". " + words.get(length - i));
        }
    }

    public static void main(String args[]) {
        letters = "abcdefg";
        List<String> words = generateWords();
        words = sortWords(words);
        displayWords(words);
    }
}

File: c#/Solver.cs
using System;
using System.IO;
using System.Collections.Generic;

class Solver {
    const string letters = "abcdefg";
    static List<string> words = new List<string>();

    static bool Valid(string word, string letters) {
        foreach (char letter in word) {
            if (word.Split(letter).Length > letters.Split(letter).Length) {
                return false;
            }
        }
        return true;
    }

    static void GenerateWords() { 
        foreach (string line in File.ReadLines("dictionary.txt"))
            if (Valid(line, letters)) {
                words.Add(line);
            }
        }
    }

    static Dictionary<char, int> scores = new Dictionary<char, int> {
        {'a', 1}, {'b', 3}, {'c', 3}, {'d', 2},
        {'e', 1}, {'f', 4}, {'g', 2}, {'h', 4},
        {'i', 1}, {'j', 8}, {'k', 5}, {'l', 1},
        {'m', 3}, {'n', 1}, {'o', 1}, {'p', 3},
        {'q', 10},{'r', 1}, {'s', 1}, {'t', 1},
        {'u', 1}, {'v', 4}, {'w', 4}, {'x', 8},
        {'y', 4}, {'z', 10}
    };

    static int Score(string word) {
        int total = 0;
        foreach (char letter in word) 
            total += scores[letter];
        
        return total;
    }

    class Comparer : IComparer<string> {
        public int Compare(string x, string y) {
            return Score(x) - Score(y);
        }
    }

    static void DisplayWords() {
        for (int i = words.Count; i > 0; i--) {
            int index = words.Count - i;
            Console.WriteLine(String.Format("{0}. {1}", i, words[index]));
        }
    }

    public static void Main(string[] args) {
        GenerateWords();
        words.Sort(new Comparer());
        DisplayWords();
    }
}
